File: old/havok.jl
using LinearAlgebra

# Function for obtain Hankel Matrix of time-delay embeddings
function Hankel(z, rows)
    cols = length(z) - rows + 1
    H = zeros(rows, cols)

    for i ∈ 1:rows
        H[i,:] .= z[i:i+cols - 1]
    end

    return H
end



# assume Zs is n rows × k features
function Hankel_nd(Zs, n_embedding)
    return hcat([vcat(Zs'[:, i-n_embedding+1:i]...) for i ∈ n_embedding:size(Zs, 1)]...)
end




function true_polys(rows, Δt, r, center=false)
    m = rows ÷ 2  # int division
    p = range(-m*Δt, stop=m*Δt, length=rows)

    U = []
    for j ∈ 1:r
        if center
            push!(U, p .^ (j))
        else
            push!(U, p .^ (j-1))
        end
    end
    U = hcat(U...)

    # Gram-Schmidt to create ONB

    # 1. normalize u₁ to obtain ê₁
    # 2. e₂ = u₂ - (ê₁⋅ u₂)u₂ then normalize to obtain ê₂
    # 3. ...repeat via projection for rest of vectors
    Q = zeros(rows, r)
    for j ∈ 1:r
        v = U[:,j]
        for k ∈ 1:(j-1)
            r_jk = dot(Q[:,k], U[:,j])
            v -= (r_jk * Q[:,k])
        end
        r_jj = norm(v)
        Q[:,j] = v / r_jj
    end

    return Q
end



function HAVOK(H, Δt, r, norm)
    U,σ,V = svd(H)

    # since we know the off-diagonals
    # should be anti-symmetric
    # go through matrix and flip sign
    # using the theoreticl basis

    polys = true_polys(size(H,1), Δt, r)
    for j ∈ 1:r
        if (dot(U[:,j], polys[:,j]) < 0)
            U[:,j] *= -1
            V[:,j] *= -1
        end
    end

    V₁ = V[1:end-1,1:r]
    V₂ = V[2:end,1:r]

    return (V₂'*V₁ - I)/(Δt * norm), U, σ, V
end


# V₁ = V₁[:, 1:r]
# V₂ = V₂[:, 1:r]

# v̇ = Av
# V̇' ≈ (V₂' - V₁')/Δt
#
# use forward Euler
#
# AV₁' = (V₂' - V₁')/Δt
#
# Note: V₁,V₂ are orthogonal
# A = (V₂' - V₁')V₁/(Δt)
#   = (V₂'V₁ - I)/(Δt)
function sHAVOK(H, Δt, r, norm)
    H₁ = H[:, 1:end-1]
    H₂ = H[:, 2:end]

    U₁, σ₁, V₁ = svd(H₁)
    U₂, σ₂, V₂ = svd(H₂)

    # since we know the off-diagonals
    # should be anti-symmetric
    # go through matrix and flip sign
    # using the theoreticl basis
    polys = true_polys(size(H,1), Δt, r)
    for j ∈ 1:r
        if (dot(U₁[:,j], polys[:,j]) < 0)
            U₁[:,j] *= -1
            V₁[:,j] *= -1
        end

        if (dot(U₂[:,j], polys[:,j]) < 0)
            U₂[:,j] *= -1
            V₂[:,j] *= -1
        end
    end

    Vr = V₁[:,1:r]
    Ur = U₁[:,1:r]
    σr = σ₁[1:r]

    return ((V₂'*V₁)[1:r,1:r] - I) / (Δt * norm), Ur, σr, Vr
end


# function sHAVOK_multi(Hs, Δt, r, norm)
#     H1s = [H[:, 1:end-1] for H ∈ Hs]
#     H2s = [H[:, 2:end] for H ∈ Hs]

#     H₁ = hcat(H1s...)
#     H₂ = hcat(H2s...)

#     U₁, σ₁, V₁ = svd(H₁)
#     U₂, σ₂, V₂ = svd(H₂)

#     # since we know the off-diagonals
#     # should be anti-symmetric
#     # go through matrix and flip sign
#     # using the theoreticl basis
#     polys = true_polys(size(H₁,1), Δt, r)
#     for j ∈ 1:r
#         if (dot(U₁[:,j], polys[:,j]) < 0)
#             U₁[:,j] *= -1
#             V₁[:,j] *= -1
#         end

#         if (dot(U₂[:,j], polys[:,j]) < 0)
#             U₂[:,j] *= -1
#             V₂[:,j] *= -1
#         end
#     end

#     Vr = V₁[:,1:r]
#     Ur = U₁[:,1:r]
#     σr = σ₁[1:r]

#     return ((V₂'*V₁)[1:r,1:r] - I) / (Δt * norm), Ur, σr
# end





function sHAVOK_central(H, Δt, r, norm)
    H₁ = H[:, 1:end-2]
    H₂ = H[:, 2:end-1]
    H₃ = H[:, 3:end]

    U₁, σ₁, V₁ = svd(H₁)
    U₂, σ₂, V₂ = svd(H₂)
    U₃, σ₃, V₃ = svd(H₃)

    polys = true_polys(size(H,1), Δt, r)
    for j ∈ 1:r
        if (dot(U₁[:,j], polys[:,j]) < 0)
            U₁[:,j] *= -1
            V₁[:,j] *= -1
        end

        if (dot(U₂[:,j], polys[:,j]) < 0)
            U₂[:,j] *= -1
            V₂[:,j] *= -1
        end

        if (dot(U₃[:,j], polys[:,j]) < 0)
            U₃[:,j] *= -1
            V₃[:,j] *= -1
        end
    end

    Vr = V₂[:,1:r]
    Ur = U₂[:,1:r]
    σr = σ₂[1:r]

    return ((V₃'*V₂)[1:r,1:r] - (V₁'*V₂)[1:r,1:r]) / (Δt * norm), Ur, σr, Vr
end





# solve using Matrix Exponential for each step.
# a single step is given via matrix exponential
#
# | v_next |       | A⋅Δt   B⋅Δt | |    v_now       |
# | f_next | = exp |  0      0  | |    f_now       |
#
#
function make_expM_const(A, B, Δt, r_model, n_control)
    r = r_model + n_control
    M = zeros(r,r)

    M[1:r_model, 1:r_model] .= (A * Δt)
    M[1:r_model, r_model+1:end] .= (B * Δt)
    expM = exp(M)

    expA = expM[1:r_model, 1:r_model]
    expB = expM[1:r_model, r_model+1:end]

    return expA, expB
end



function step_const!(v_next, v_now, f_now, expA, expB)
    v_next .= expA*v_now + expB*f_now
end



# solve using Matrix Exponential for each step.
# a single step is given via matrix exponential
#
# |     v_next     |       | A⋅Δt   B⋅Δt  0 | |    v_now       |
# |     f_next     | = exp |  0      0   I | |    f_now       |
# | f_next - f_now | = exp |  0      0   0 | | f_next - f_now |
#
#
function make_expM_linear(A, B, Δt, r_model, n_control)
    r = r_model + n_control
    M = zeros(r+n_control,r+n_control)

    M[1:r_model, 1:r_model] .= (A * Δt)      # r_model × r_model
    M[1:r_model, r_model+1:r] .= (B * Δt)    # r_model × n_control
    M[r_model+1:r, r+1:end] .= I(n_control)  # r_model × n_control
    expM = exp(M)

    expA = expM[1:r_model, 1:r_model]
    expB = expM[1:r_model, r_model+1:r]
    exp0 = expM[1:r_model, r+1:end]
    return expA, expB, exp0
end


function step_linear!(v_next, v_now, f_next, f_now, expA, expB, exp0)
    v_next .= expA*v_now + expB*f_now + exp0*(f_next .- f_now)
end



function eval_havok(Zs, ts, n_embedding, r_model, n_control)
    r = r_model + n_control

    # construct Hankel Matrix
    H = Hankel(Zs, n_embedding)

    # cutoff time for training vs testing partition
    dt = ts[2]-ts[1]
    Zs_x = H[end, :]
    ts_x = range(ts[n_embedding], step=dt, length=length(Zs_x))

    # compute havok decomposition
    # Ξ,U,σ,V = sHAVOK(H, dt, r, 1);
    Ξ,U,σ,_ = sHAVOK(H, dt, r, 1);

    V = H'*U*Diagonal(1 ./ σ)

    # further truncate original time series to match
    # the data in V
    # Zs_x = Zs_x[1:end-1]
    # ts_x = ts_x[1:end-1]
    Zs_x = Zs_x[1:end]
    ts_x = ts_x[1:end]


    # select Linear and Forcing coef. Matrices
    A = Ξ[1:r_model, 1:r_model];
    B = Ξ[1:r_model, r_model+1:end];

    # set up initial condition
    v₁ = V[1,1:r_model]

    # pick out forcing values
    fvals = V[:,r_model+1:r]

    # construct exponential matrices for time evolution
    expA, expB = make_expM_const(A, B, dt, r_model, n_control)

    # set up outgoing array
    Vout = zeros(size(V, 1), r_model);
    Vout[1,:] .= v₁;
    v_tmp = similar(v₁);

    # compute time evolution
    for i ∈ 2:size(Vout,1)
        step_const!(v_tmp, Vout[i-1,:], fvals[i-1,:], expA, expB)
        Vout[i,:] .= v_tmp
    end

    # reconstruct original time series
    # Ĥ = U*Diagonal(σ)*hcat(Vout, fvals)'
    Ĥ = U[:,1:r_model]*Diagonal(σ[1:r_model])*Vout'
    Ẑs_x = Ĥ[end,:]

    return Zs_x, Ẑs_x, ts_x, U, σ, Vout, A, B, fvals
end


function integrate_havok(Zs, ts, n_embedding, r_model, n_control, A, B, U, σ)
    r = r_model + n_control

    # construct Hankel Matrix
    H = Hankel(Zs, n_embedding)

    # get the current V matrix using U,σ
    # H = UΣV'
    # Σ⁻¹U'H = V'
    # V = H'UΣ⁻¹
    Vcur = H'*U*Diagonal(1 ./ σ)

    # Generate time series via Hankel Matrix
    dt = ts[2]-ts[1]
    Zs_x = H[end, :]
    ts_x = range(ts[n_embedding], step=dt, length=length(Zs_x))

    # set up initial condition
    v₁ = Vcur[1,1:r_model]

    # pick out forcing values
    fvals = Vcur[:,r_model+1:r]


    # construct exponential matrices for time evolution
    expA, expB = make_expM_const(A, B, dt, r_model, n_control)

    # set up outgoing array
    Vout = zeros(size(Vcur, 1), r_model);
    Vout[1,:] .= v₁;
    v_tmp = similar(v₁);

    # compute time evolution
    for i ∈ 2:size(Vout,1)
        step_const!(v_tmp, Vout[i-1,:], fvals[i-1,:], expA, expB)
        Vout[i,:] .= v_tmp
    end

    # reconstruct original time series
    # Ĥ = U*Diagonal(σ)*hcat(Vout, fvals)'
    Ĥ = U*Diagonal(σ)*hcat(Vout, fvals)'
    Ĥ = U[:,1:r_model]*Diagonal(σ[1:r_model])*Vout'
    Ẑs_x = Ĥ[end,:]

    return Zs_x, Ẑs_x, ts_x
end



get_v_from_z(z, invΣ, U) = invΣ*U'*z
get_z_from_v(v, Σ, U, r_model) = U[:, 1:r_model]*Σ[1:r_model, 1:r_model]*v[1:r_model]
get_z_from_v_and_f(v, f, Σ, U, r) = U[:, 1:r]*Σ[1:r, 1:r]*vcat(v,f)



using Random

function forcing_model(fvals, n_embedding, n_control)
    fs_emb = Hankel_nd(fvals, n_embedding)

    fs = fs_emb[end - n_control + 1 : end,:]

    F = fs_emb[:, 1:end-1]
    f_next = fs[:, 2:end]

    # verify we have shifted things correctly
    @assert all(F[end - n_control + 1,2] .== f_next[:,1])


    size(f_next)
    ntrain = Int(0.8*size(f_next,2))
    idx_train = shuffle(1:size(f_next,2))[1:ntrain]
    idx_test = setdiff(1:size(f_next,2), idx_train)

    Ftrain = F[:, idx_train]
    Ftest = F[:, idx_test]
    f_train = f_next[:, idx_train]
    f_test = f_next[:, idx_test]

    # M = f_train / vcat(Ftrain, ones(size(f_train)))

    K = f_train / Ftrain

    # K = M[:, 1:end-1]
    # b = M[:, end]

    # now create predictions for forcing function
    # f̂_train = (K*Ftrain .+ b)
    # f̂_test = (K*Ftest .+ b)

    f̂_train = K*Ftrain
    f̂_test = K*Ftest

    # return K, b, f_train, f̂_train, f_test, f̂_test
    return K, f_train, f̂_train, f_test, f̂_test
end



function get_K_for_forcing(Zs, fvals, n_embedding)
    Zvec = Hankel(Zs_train, n_embedding)
    Zs_x
    size(Zvec)
    size(fvals)

    Zvec_now = Zvec[:, 1:end-1]
    z_next = Zs_x[2:end]
    fnow = fvals[1:end-1,:]
    fnext = fvals[2:end,:]

    # now we want to fit a model to map our
    # fnext = K * vcat(fnow, z_next, Zvec_now)
    K = fnext' / vcat(fnow', z_next', Zvec_now)

    # evaluate
    f̂next = (K * (vcat(fnow', z_next', Zvec_now)))'

    return K, fnext, f̂next
end



# function forecast_havok(Zs, ts, n_embedding, r_model, n_control, A, B, U, σ, K)
#     dt = ts[2]-ts[1]
#     r = r_model + n_control

#     # set up outgoing array for time series predictions
#     Zs_out = Zs[n_embedding:end]
#     Ẑs_out = zeros(length(Zs[n_embedding:end]))
#     ts_out = range(ts[n_embedding], step=dt, length=length(Zs_out))

#     # current time series embedding
#     Zvec = Zs[1:n_embedding]

#     # get first embedding vector for time delay
#     V_0 = Diagonal(1 ./ σ)*U'*Zs[1:n_embedding]

#     # split into initial state and forcing vectors
#     vnow = V_0[1:r_model]
#     fnow = V_0[r_model+1:r]

#     fnext_true = (Diagonal(1 ./ σ)*U'*Zs[2:n_embedding+1])[r_model+1:r]

#     # construct exponential matrices for time evolution
#     expA, expB = make_expM_const(A, B, dt, r_model, n_control)

#     # set up vectors for predicted values
#     vnext = similar(vnow);      # for updating state
#     fnext= similar(fnow);      # for updating forcing

#     # pre-compute this for re-use
#     σu = σ[1:r_model] .* U[end,1:r_model]
#     # invΣUt = Diagonal(1 ./ σ[r_model+1:r]) * U[:, r_model+1:r]'

#     # compute time evolution
#     for i ∈ 1:length(Zs_out)

#         # i=1

#         # integrate state forward
#         step_const!(vnext, vnow, fnow, expA, expB)

#         # update the state
#         # vnext .= vnow
#         vnow .= vnext

#         # convert state back to time series value
#         znext = dot(σu, vnext)
#         Ẑs_out[i] = znext

#         Ẑs_out[1]
#         Zs_out[1]

#         # use K to update forcing
#         fnext = K*vcat(fnow, znext, Zvec)
#         fnow = fnext


#         # update embedding vector
#         Zvec[1:end-1] .= Zvec[2:end]
#         Zvec[end] = znext
#     end

#     return Zs_out, Ẑs_out, ts_out
# end


function forecast_havok(Zs, ts, n_embedding, r_model, n_control, A, B, U, σ)
    dt = ts[2]-ts[1]
    r = r_model + n_control

    # set up outgoing array for time series predictions
    Zs_out = Zs[n_embedding:end]
    Ẑs_out = zeros(length(Zs[n_embedding:end]))
    ts_out = range(ts[n_embedding], step=dt, length=length(Zs_out))

    # current time series embedding
    Zvec = Zs[1:n_embedding]

    # get first embedding vector for time delay
    V_0 = Diagonal(1 ./ σ)*U'*Zs[1:n_embedding]

    # split into initial state and forcing vectors
    vnow = V_0[1:r_model]
    fnow = V_0[r_model+1:r]

    # construct exponential matrices for time evolution
    expA, expB = make_expM_const(A, B, dt, r_model, n_control)

    # set up vectors for predicted values
    vnext = similar(vnow);      # for updating state
    fnext= similar(fnow);      # for updating forcing

    # pre-compute this for re-use
    σu = σ[1:r_model] .* U[end,1:r_model]
    invΣUt = Diagonal(1 ./ σ[r_model+1:end]) * U[:, r_model+1:end]'

    # compute time evolution
    for i ∈ 1:length(Zs_out)

        # i=1

        # integrate state forward
        step_const!(vnext, vnow, fnow, expA, expB)

        # update the state
        # vnext .= vnow
        vnow .= vnext

        # convert state back to time series value
        znext = dot(σu, vnext)
        Ẑs_out[i] = znext

        # update embedding vector
        Zvec[1:end-1] .= Zvec[2:end]
        Zvec[end] = znext

        # update forcing
        fnext = invΣUt*Zvec
        fnow = fnext

        # Ẑs_out[1]
        # Zs_out[1]

        # # use K to update forcing
        # fnext = K*vcat(fnow, znext, Zvec)
        # fnow = fnext


        # # update embedding vector
        # Zvec[1:end-1] .= Zvec[2:end]
        # Zvec[end] = znext

    end

    return Zs_out, Ẑs_out, ts_out
end

File: havok.py
import numpy as np
from scipy.linalg import expm
from scipy.signal import lsim
from tqdm import tqdm

# rows is the embedding size
def build_hankel(z, rows):
    cols = len(z) - rows + 1
    H = np.empty((rows, cols))

    for i in range(rows):
        H[i,:] = z[i:i+cols]
    return H

def build_hankel_multi(Zs, rows):
    return np.hstack([build_hankel(z, rows) for z in Zs])

def true_polys(rows, dt, r, center=False): 
    m = rows // 2

    p = np.linspace(-m*dt, m*dt, rows)
    U = []
    for j in range(r):
        if (center):
            U.append(p ** (j + 1))
        else: 
            U.append(p ** j)
    U = np.vstack(U).T

    Q = np.empty((rows, r)) 
    
    # Perform Gram-Schmidt to yield ONB

    # 1. normalize u1 to obtain e1_hat
    # 2. e2 = u2 - (e1_hat * u2)u2 then normalize to obtain e2_hat
    # 3. ... repeat via projection for rest of vectors
    for j in range(r): 
        v = U[:, j]
        for k in range(j - 1): 
            # r_jk = Q[:, k].T @ U[:, j]
            r_jk = np.dot(Q[:, k], U[:, j])
            v -= (r_jk * Q[:, k])
        r_jj = np.linalg.norm(v)
        Q[:, j] = v / r_jj
    return Q


def HAVOK(H, dt, r): 
    U, s, Vh = np.linalg.svd(H, full_matrices=False)
    V = Vh.T

    # use polynomials to correct off diagonals
    # to enforce anti-symmetry
    polys = true_polys(H.shape[0], dt, r)
    for j in range(r): 
        # if our vector and the poly are anti-aligned
        # then re-align
        if (np.dot(U[:,j], polys[:,j]) < 0.0):
            U[:, j] *= -1
            V[:, j] *= -1

    # Split up embeddings for difference computation
    V1 = V[:-1, :r]
    V2 = V[1:, :r]

    # note: we need to use pinv since slices of V may
    # not be unitary
    A = (V2.T @ np.linalg.pinv(V1.T) - np.eye(r)) / dt
    return A, U[:, :r], s[:r], V[:, :r]


def sHAVOK(H, dt, r):
    # instead of splitting V after SVD
    # we split H before SVD to maintain
    # unitarity of V slices
    H1 = H[:, :-1]
    H2 = H[:, 1:]

    U1, s1, Vh1 = np.linalg.svd(H1, full_matrices=False)
    U2, s2, Vh2 = np.linalg.svd(H2, full_matrices=False)

    # create views of V w/o adjoint
    V1 = Vh1.T 
    V2 = Vh2.T

    polys = true_polys(H.shape[0], dt, r)
    for j in range(r):
        if (np.dot(U1[:,j], polys[:,j]) < 0.0):
            U1[:,j] *= -1
            V1[:,j] *= -1        
        if (np.dot(U2[:,j], polys[:,j]) < 0.0):
            U2[:,j] *= -1
            V2[:,j] *= -1

    # now we don't need pinv sinve V1 and V2 are unitary
    # interestingly here we are only truncating *after*
    # the multiplication... will that have an impact?
    A = ((V2.T @ V1)[:r,:r] - np.eye(r)) / dt


    return A, U1[:, :r], s1[:r], V1[:, :r]


def make_expM(A, B, dt, r_model, n_control):
    r = r_model + n_control
    M = np.zeros((r,r))
    M[:r_model, :r_model] = A * dt
    M[:r_model, r_model:r] = B * dt
    expM = expm(M)

    expA = expM[:r_model, :r_model]
    expB = expM[:r_model, r_model:r]

    return expA, expB

# def take_step(v_now, f_now, expA, expB):
#     return expA @ v_now + expB @ f_now


def eval_havok(z, ts, n_embedding, r_model, n_control):
    r = r_model + n_control

    # construct Hankel matrix
    #print("...building Hankel matrix")
    H = build_hankel(z, n_embedding)

    # get true values of time series
    # after forming Hankel matrix
    dt = ts[1] - ts[0]
    z_x = H[-1, :]
    t_x = np.arange(ts[n_embedding], ts[n_embedding] + dt*len(z_x),step=dt)

    # compute sHAVOK decomp
    # M, U, s, _ = sHAVOK(H, dt, r)
    #print("...fitting HAVOK model")
    M, U, s, _ = sHAVOK(H, dt, r)

    # using our H, U, and s, reconstruct the dembeddings
    # V = H.T @ (U * (1 / s)[np.newaxis, :])
    V = H.T @ U @ np.diag(1.0 / s)

    # pick out the A and B matrices
    A = M[:r_model, :r_model]
    B = M[:r_model, r_model:r]


    # pick out forcing values
    fvals = V[:, r_model:r]

    # construct exp matrices for integration
    expA, expB = make_expM(A, B, dt, r_model, n_control)

    # set up outgoing arrays
    Vout = np.zeros((V.shape[0], r_model))

    # set initial condition
    Vout[0,:] = V[0, :r_model]

    # time-evolve the system
    #print(f"...integrating from t={t_x[0]:.3f} to t={t_x[-1]:.3f}")
    for i in range(Vout.shape[0]-1):
        Vout[i+1,:] = expA @ Vout[i,:] + expB @ fvals[i,:]

    # reconstruct Hankel matrix and get time series
    H_pred = U[:, :r_model] @ np.diag(s[:r_model]) @ Vout.T    
    z_pred = H_pred[-1,:]

    return z_x, z_pred, t_x, U, s, Vout, A, B, fvals





def eval_havok_multi(zs, ts, n_embedding, r_model, n_control):
    r = r_model + n_control

    # construct Hankel matrix
    # print("...building Hankel matrices")
    Hs = []
    dts = []
    zs_x = []
    ts_x = []
    # for i in tqdm(range(len(zs))):
    for i in range(len(zs)):
        H = build_hankel(zs[i], n_embedding)
        Hs.append(H)

        dt = ts[i][1] - ts[i][0]
        dts.append(dt)
        
        z_x = H[-1,:]
        # t_x = np.arange(ts[i][n_embedding], ts[i][n_embedding] + dt*len(z_x), step=dt)
        t_x = ts[i][n_embedding-1:]

        zs_x.append(z_x)
        ts_x.append(t_x)

    H = np.hstack(Hs)        

    assert np.allclose(dts, dts[0])
    dt = dts[0]

    # compute sHAVOK decomp
    # print("...fitting HAVOK model")
    M, U, s, _ = sHAVOK(H, dt, r)

    # using our H, U, and s, reconstruct the dembeddings
    Vs = [H.T @ U @ np.diag(1.0 / s) for H in Hs]

    # pick out forcing values
    fvals = [V[:, r_model:r] for V in Vs]

          # set up predictded embedding matrices
    Vs_out = [np.zeros((V.shape[0], r_model)) for V in Vs]

    # set initial condition
    for i in range(len(Vs)):
        Vs_out[i][0,:] = Vs[i][0,:r_model]


    # pick out the A and B matrices
    A = M[:r_model, :r_model]
    B = M[:r_model, r_model:r]


    # construct exp matrices for integration
    expA, expB = make_expM(A, B, dt, r_model, n_control)


    # time-evolve the system
    zs_pred = []

    # for k in tqdm(range(len(ts_x))):
    for k in range(len(ts_x)):
        # perform integration
        for i in range(Vs_out[k].shape[0]-1):
            Vs_out[k][i+1,:] = expA @ Vs_out[k][i,:] + expB @ fvals[k][i,:]

        H_pred = U[:, :r_model] @ np.diag(s[:r_model]) @ Vs_out[k].T
        zs_pred.append(H_pred[-1,:])


    return zs_x, zs_pred, ts_x, U, s, Vs_out, A, B, fvals




# Assuming you already have a model, integrate the existing
# model for new time series points (t, z(t))
# this is useful for checking performance on a test set
def integrate_havok(z, t, n_embedding, r_model, n_control, A, B, U, s):
    r = r_model + n_control

    H = build_hankel(z, n_embedding)
    V = H.T @ U @ np.diag(1.0 / s)

    # pick out forcing values
    fvals = V[:, r_model:r]

    # compute dt
    dt = t[1] - t[0]
    z_x = H[-1, :]
    t_x = t[n_embedding-1:]


    # construct exp matrices for integration
    expA, expB = make_expM(A, B, dt, r_model, n_control)

    # set up outgoing arrays
    Vout = np.zeros((V.shape[0], r_model))

    # set initial condition
    Vout[0,:] = V[0, :r_model]

    # time-evolve the system
    # print(f"...integrating from t={t_x[0]:.3f} to t={t_x[-1]:.3f}")
    for i in range(Vout.shape[0]-1):
        Vout[i+1,:] = expA @ Vout[i,:] + expB @ fvals[i,:]

    # reconstruct Hankel matrix and get time series
    H_pred = U[:, :r_model] @ np.diag(s[:r_model]) @ Vout.T    
    z_pred = H_pred[-1,:]

    return z_x, z_pred, t_x

